`match` doesn't check each branch produces a value

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relvant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** recent git master
**OS:** Windows 8

**What did you do?**
```v
_ = match true {
	true {1}
	else {}
}

_ = match true {
	true {}
	else {1}
}
```

**What did you expect to see?**
E.g.
```
Error line 3: else branch does not produce a value.
Error line 7: true branch does not produce a value.
```
 
**What did you see instead?**
```
==================
C:\Users\CZASTK~1\AppData\Local\Temp\v\matchvals.tmp.c: In function 'main__main':
C:\Users\CZASTK~1\AppData\Local\Temp\v\matchvals.tmp.c:8003:40: error: expected expression before ')' token
   {any_int _ = (_t75 == true) ? (1) : ();}
                                        ^
C:\Users\CZASTK~1\AppData\Local\Temp\v\matchvals.tmp.c:8006:34: error: expected expression before ')' token
   {any_int _ = (_t76 == true) ? () : (1);}
                                  ^
...
==================
(Use `v -cg` to print the entire error message)
```

Similar error

71d3d4c

```v
type Arr = []string | []int

fn (arr Arr) str() string {
        return match arr {
                []string {
                        arr.join(' ')
                }
                []int {
                        return arr.str()
                }
        }
}

fn main() {
        println(Arr([0, 0]))
        println(Arr(['0', '0']))
}
```

```
==================
                                                                                                                   ^
/tmp/v/match.13114379231758164442.tmp.c:8441:2: error: expected expression
        ));
        ^
2 errors generated.
...
==================
(Use `v -cg` to print the entire error message)

```

--- a/vlib/v/ast/ast.v
+++ b/vlib/v/ast/ast.v
@@ -773,7 +773,8 @@ pub:
 	stmts         []Stmt      // right side
 	pos           token.Position
 	is_else       bool
-	post_comments []Comment // comments below ´... }´
+	post_comments []Comment      // comments below ´... }´
+	branch_pos    token.Position // for checker errors about invalid branches
 pub mut:
 	exprs []Expr // left side
 	scope &Scope

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -5517,7 +5517,7 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 				}
 			} else if ret_type != ast.void_type {
 				c.error('`match` expression requires an expression as the last statement of every branch',
-					branch.pos)
+					branch.branch_pos)
 			}
 		}
 		// If the last statement is an expression, return its type

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -5556,7 +5556,7 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 				}
 			}
 		}
-		
+
 		first_iteration = false
 		if has_return := c.has_return(branch.stmts) {
 			if has_return {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -5516,7 +5516,8 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 					}
 				}
 			} else if ret_type != ast.void_type {
-				c.error('missing branch in `match` expression', branch.pos)
+				c.error('`match` expression requires an expression as the last statement of every branch',
+					branch.pos)
 			}
 		}
 		// If the last statement is an expression, return its type
@@ -5548,7 +5549,7 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 				}
 				else {
 					if node.is_expr && ret_type != ast.void_type {
-						c.error('statement cannot be used as return type `${c.table.get_type_symbol(ret_type).name}` in `match` expression',
+						c.error('`match` expression requires an expression as the last statement of every branch',
 							stmt.pos)
 					}
 				}

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -5547,16 +5547,13 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 					}
 				}
 				else {
-					// TODO: ask alex about this
-					// typ := c.expr(stmt.expr)
-					// type_sym := c.table.get_type_symbol(typ)
-					// p.warn('match expr ret $type_sym.name')
-					// node.typ = typ
-					// return typ
+					if node.is_expr && ret_type != ast.void_type {
+						c.error('statement cannot be used as return type `${c.table.get_type_symbol(ret_type).name}` in `match` expression',
+							stmt.pos)
+					}
 				}
 			}
 		}
-
 		first_iteration = false
 		if has_return := c.has_return(branch.stmts) {
 			if has_return {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -5502,14 +5502,20 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 	mut nbranches_without_return := 0
 	for branch in node.branches {
 		c.stmts(branch.stmts)
-		if node.is_expr && branch.stmts.len > 0 {
-			// ignore last statement - workaround
-			// currently the last statement in a match branch does not have an
-			// expected value set, so e.g. IfExpr.is_expr is not set.
-			// probably any mismatch will be caught by not producing a value instead
-			for st in branch.stmts[0..branch.stmts.len - 1] {
-				// must not contain C statements
-				st.check_c_expr() or { c.error('`match` expression branch has $err.msg', st.pos) }
+		if node.is_expr {
+			if branch.stmts.len > 0 {
+				// ignore last statement - workaround
+				// currently the last statement in a match branch does not have an
+				// expected value set, so e.g. IfExpr.is_expr is not set.
+				// probably any mismatch will be caught by not producing a value instead
+				for st in branch.stmts[0..branch.stmts.len - 1] {
+					// must not contain C statements
+					st.check_c_expr() or {
+						c.error('`match` expression branch has $err.msg', st.pos)
+					}
+				}
+			} else {
+				c.error('empty branch in `match` expression', branch.pos)
 			}
 		}
 		// If the last statement is an expression, return its type

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -5497,6 +5497,7 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 	node.is_sum_type = cond_type_sym.kind in [.interface_, .sum_type]
 	c.match_exprs(mut node, cond_type_sym)
 	c.expected_type = cond_type
+	mut first_iteration := true
 	mut ret_type := ast.void_type
 	mut nbranches_with_return := 0
 	mut nbranches_without_return := 0
@@ -5514,7 +5515,7 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 						c.error('`match` expression branch has $err.msg', st.pos)
 					}
 				}
-			} else if ret_type != ast.void_type { 
+			} else if ret_type != ast.void_type {
 				c.error('missing branch in `match` expression', branch.pos)
 			}
 		}
@@ -5527,7 +5528,7 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 						c.expected_type = node.expected_type
 					}
 					expr_type := c.expr(stmt.expr)
-					if ret_type == ast.void_type {
+					if first_iteration {
 						if node.is_expr && !node.expected_type.has_flag(.optional)
 							&& c.table.get_type_symbol(node.expected_type).kind == .sum_type {
 							ret_type = node.expected_type
@@ -5554,7 +5555,9 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 					// return typ
 				}
 			}
-		} 
+		}
+		
+		first_iteration = false
 		if has_return := c.has_return(branch.stmts) {
 			if has_return {
 				nbranches_with_return++

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -5514,8 +5514,8 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 						c.error('`match` expression branch has $err.msg', st.pos)
 					}
 				}
-			} else {
-				c.error('empty branch in `match` expression', branch.pos)
+			} else if ret_type != ast.void_type { 
+				c.error('missing branch in `match` expression', branch.pos)
 			}
 		}
 		// If the last statement is an expression, return its type
@@ -5554,7 +5554,7 @@ pub fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 					// return typ
 				}
 			}
-		}
+		} 
 		if has_return := c.has_return(branch.stmts) {
 			if has_return {
 				nbranches_with_return++

--- a/vlib/v/checker/tests/match_expr_empty_branch.out
+++ b/vlib/v/checker/tests/match_expr_empty_branch.out
@@ -1,0 +1,6 @@
+vlib/v/checker/tests/match_expr_empty_branch.vv:2:2: error: empty branch in `match` expression
+    1 | _ := match true {
+    2 |     true {}
+      |     ~~~~
+    3 |     false { 0 }
+    4 | }

--- a/vlib/v/checker/tests/match_expr_empty_branch.out
+++ b/vlib/v/checker/tests/match_expr_empty_branch.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/match_expr_empty_branch.vv:3:2: error: missing branch in `match` expression
+vlib/v/checker/tests/match_expr_empty_branch.vv:3:2: error: `match` expression requires an expression as the last statement of every branch
     1 | _ := match true {
     2 |     true { 0 }
     3 |     false {}

--- a/vlib/v/checker/tests/match_expr_empty_branch.out
+++ b/vlib/v/checker/tests/match_expr_empty_branch.out
@@ -1,6 +1,6 @@
-vlib/v/checker/tests/match_expr_empty_branch.vv:2:2: error: empty branch in `match` expression
+vlib/v/checker/tests/match_expr_empty_branch.vv:3:2: error: missing branch in `match` expression
     1 | _ := match true {
-    2 |     true {}
-      |     ~~~~
-    3 |     false { 0 }
+    2 |     true { 0 }
+    3 |     false { }
+      |     ~~~~~
     4 | }

--- a/vlib/v/checker/tests/match_expr_empty_branch.out
+++ b/vlib/v/checker/tests/match_expr_empty_branch.out
@@ -1,6 +1,6 @@
 vlib/v/checker/tests/match_expr_empty_branch.vv:3:2: error: missing branch in `match` expression
     1 | _ := match true {
     2 |     true { 0 }
-    3 |     false { }
-      |     ~~~~~
+    3 |     false {   }
+      |     ~~~~~~~~~~~
     4 | }

--- a/vlib/v/checker/tests/match_expr_empty_branch.out
+++ b/vlib/v/checker/tests/match_expr_empty_branch.out
@@ -1,6 +1,6 @@
 vlib/v/checker/tests/match_expr_empty_branch.vv:3:2: error: missing branch in `match` expression
     1 | _ := match true {
     2 |     true { 0 }
-    3 |     false {   }
-      |     ~~~~~~~~~~~
+    3 |     false {}
+      |     ~~~~~~~~
     4 | }

--- a/vlib/v/checker/tests/match_expr_empty_branch.vv
+++ b/vlib/v/checker/tests/match_expr_empty_branch.vv
@@ -1,4 +1,4 @@
 _ := match true {
 	true { 0 }
-	false {   }
+	false {}
 }

--- a/vlib/v/checker/tests/match_expr_empty_branch.vv
+++ b/vlib/v/checker/tests/match_expr_empty_branch.vv
@@ -1,4 +1,4 @@
 _ := match true {
-	true {}
+	true { }
 	false { 0 }
 }

--- a/vlib/v/checker/tests/match_expr_empty_branch.vv
+++ b/vlib/v/checker/tests/match_expr_empty_branch.vv
@@ -1,4 +1,4 @@
 _ := match true {
 	true { 0 }
-	false { }
+	false {}
 }

--- a/vlib/v/checker/tests/match_expr_empty_branch.vv
+++ b/vlib/v/checker/tests/match_expr_empty_branch.vv
@@ -1,0 +1,4 @@
+_ := match true {
+	true {}
+	false { 0 }
+}

--- a/vlib/v/checker/tests/match_expr_empty_branch.vv
+++ b/vlib/v/checker/tests/match_expr_empty_branch.vv
@@ -1,4 +1,4 @@
 _ := match true {
-	true { }
-	false { 0 }
+	true { 0 }
+	false { }
 }

--- a/vlib/v/checker/tests/match_expr_empty_branch.vv
+++ b/vlib/v/checker/tests/match_expr_empty_branch.vv
@@ -1,4 +1,4 @@
 _ := match true {
 	true { 0 }
-	false {}
+	false {   }
 }

--- a/vlib/v/checker/tests/match_expr_non_void_stmt_last.out
+++ b/vlib/v/checker/tests/match_expr_non_void_stmt_last.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/match_expr_non_void_stmt_last.vv:9:4: error: statement cannot be used as return type `string` in `match` expression
+vlib/v/checker/tests/match_expr_non_void_stmt_last.vv:9:4: error: `match` expression requires an expression as the last statement of every branch
     7 |         }
     8 |         []int {
     9 |             return arr.str()

--- a/vlib/v/checker/tests/match_expr_non_void_stmt_last.out
+++ b/vlib/v/checker/tests/match_expr_non_void_stmt_last.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/match_expr_non_void_stmt_last.vv:9:4: error: statement cannot be used as return type `string` in `match` expression
+    7 |         }
+    8 |         []int {
+    9 |             return arr.str()
+      |             ~~~~~~~~~~~~~~~~
+   10 |         }
+   11 |     }

--- a/vlib/v/checker/tests/match_expr_non_void_stmt_last.vv
+++ b/vlib/v/checker/tests/match_expr_non_void_stmt_last.vv
@@ -1,0 +1,17 @@
+type Arr = []int | []string
+
+fn (arr Arr) str() string {
+	return match arr {
+		[]string {
+			arr.join(' ')
+		}
+		[]int {
+			return arr.str()
+		}
+	}
+}
+
+fn main() {
+	println(Arr([0, 0]))
+	println(Arr(['0', '0']))
+}

--- a/vlib/v/parser/if_match.v
+++ b/vlib/v/parser/if_match.v
@@ -229,19 +229,22 @@ fn (mut p Parser) match_expr() ast.MatchExpr {
 				p.check(.comma)
 			}
 		}
+		branch_last_pos := p.prev_tok.position()
 		// p.warn('match block')
 		p.inside_match_body = true
 		stmts := p.parse_block_no_scope(false)
 		branch_scope := p.scope
 		p.close_scope()
 		p.inside_match_body = false
-		pos := branch_first_pos.extend_with_last_line(p.tok.position(), p.tok.line_nr)
+		pos := branch_first_pos.extend_with_last_line(branch_last_pos, p.prev_tok.line_nr)
+		branch_pos := branch_first_pos.extend_with_last_line(p.tok.position(), p.tok.line_nr)
 		post_comments := p.eat_comments({})
 		branches << ast.MatchBranch{
 			exprs: exprs
 			ecmnts: ecmnts
 			stmts: stmts
 			pos: pos
+			branch_pos: branch_pos
 			is_else: is_else
 			post_comments: post_comments
 			scope: branch_scope

--- a/vlib/v/parser/if_match.v
+++ b/vlib/v/parser/if_match.v
@@ -229,14 +229,13 @@ fn (mut p Parser) match_expr() ast.MatchExpr {
 				p.check(.comma)
 			}
 		}
-		branch_last_pos := p.prev_tok.position()
 		// p.warn('match block')
 		p.inside_match_body = true
 		stmts := p.parse_block_no_scope(false)
 		branch_scope := p.scope
 		p.close_scope()
 		p.inside_match_body = false
-		pos := branch_first_pos.extend_with_last_line(branch_last_pos, p.prev_tok.line_nr)
+		pos := branch_first_pos.extend_with_last_line(p.tok.position(), p.tok.line_nr)
 		post_comments := p.eat_comments({})
 		branches << ast.MatchBranch{
 			exprs: exprs

